source_manager.go

```diff
@@ -141,15 +158,121 @@ func NewSourceManager(an ProjectAnalyzer, cachedir string) (*SourceMgr, error) {
 		}
 	}
 
-	return &SourceMgr{
+	sm := &SourceMgr{
 		cachedir: cachedir,
 		lf:       fi,
 		srcs:     make(map[string]source),
 		srcfuts:  make(map[string]*unifiedFuture),
 		an:       an,
 		dxt:      pathDeducerTrie(),
 		rootxt:   newProjectRootTrie(),
-	}, nil
+		qch:      make(chan struct{}),
+	}
+
+	return sm, nil
+}
+
+// SetUpSigHandling sets up typical signal handling for a SourceMgr. It will
+// register a signal handler to be notified on:
+//
+//  - syscall.SIGINT
+//  - syscall.SIGHUP
+//  - syscall.SIGTERM
+//  - syscall.SIGQUIT
+//  - os.Interrupt
+func SetUpSigHandling(sm *SourceMgr) {
+	sigch := make(chan os.Signal)
```

Comment:
Over-engineering, interleaving and kernels aside, this is very simple, and mandated by the os/signal API:

> Package signal will not block sending to c: the caller must ensure that c has sufficient buffer space to keep up with the expected signal rate. For a channel used for notification of just one signal value, a buffer of size 1 is sufficient

Having no buffer at all risks missing a signal altogether. Imagine ^C'ing your process and nothing happens because nobody was receiving from sigch at the moment.